### ISSUE TITLE
Resize Gameboards to work with zone sizing logic

### ISSUE THREAD

**Issue by sIKE23** [sIKE23/Mage-Wars#237]
The 3x4 boards are easy and I added in the Marble one as requested by C Savy. I resized it to be 880x660 just like the other 3x4 boards. However the other boards are a bit more of a challenge. We have a 6x4 board, a 5x4, a 4x4, a 2x3, and a 3x3 board.

The boards that are not equal in zones (i.e 5x4, 3x4) are ok to work with as square zones. As the default in game (and only) board size is rectangular in shape (1000,750) and when presented on the table it works out fine. However with the maps with equal zone numbers if you make the map square (as the zones are square) when the board is displayed on the table it is skewed to a rectangle and the zone boundaries do not match up.

Would it be hard to enhance to the defineRectangularMap() function tilesizeX and tilesizeY to better support square boards in-game?

**Comment by ACG8** [comment]
>Would it be hard to enhance to the defineRectangularMap() function tilesizeX and tilesizeY to better support square boards in-game?

Are you asking for support for non-square zones? That would be possible, but would require a lot of code rewriting.

Otherwise, I am not sure what the issue is (but I haven't been trying to implement new boards). If the boards do not have perfectly square zones, the simplest solution is to distort the images so that they do.

**Comment by sIKE23** [comment]
That I understand, but the problem is that I can not see what that
distortion equates to on the table top and since there are no visible lines
for me to track and see where the numbers land vs the graphic on the board
I have no idea of how to adjust those numbers to get things working
correctly.

I was asking about non-square but rectangler zones.

**Comment by ACG8** [comment]
I see. Even given the difficulty of figuring out how to distort maps, I think it is probably a lot easier to modify the map images than it would be to modify the code to work with maps that are not made of uniform, perfect squares. I'm not sure what the best solution is, then.

**Comment by sIKE23** [comment]
I was thinking about making some "cards" that are transparent except for a
reddish and/or yellowish crosses that are say 10 pixels wide. Then I can
see where they snap to on the zone border and then I can just change the
numbers to bring them into alignment. The other thought is that you create
a function that can draw the grids for me. I just enter in a number into a
question box and it would draw the grid for me to work out the dimensions.

**Comment by ACG8** [comment]
Does OCTGN have functions that allow you to draw on the board? That would be very useful... Or do you mean a program to generate grids in some image editor?

**Comment by sIKE23** [comment]
There is nothing in the API to draw lines that I am aware of. I was hoping
you would say I need modules xyz to do that and then I would make it happen
:)

How about the card idea? You write simple loop that places cards on all
zone borders or zone border intersections. I then make cards as proposed. I
just need to be able to see what way I need to go with the number i.e.
defineRectangularMap(4,3,*250*). I have made all of the maps square and I
know the dimensions of each zone, but once I load them in to OCTGN it does
its thing and my numbers are useless.

**Comment by ACG8** [comment]
Sadly I don't think any module will help if the API doesn't support it :(

The card idea could work. If I understand you, you would design a transparent card that is zone-sized (much like one of the Domination tiles, only transparent), and then I would write a function that places this card on each zone of a grid generated by defineRectangularMap(). Then you would use the information to size the maps?

I don't think I will have time to do this today, though.

**Comment by sIKE23** [comment]
Think this with a transparent background:


[image: Inline image 4]

You can use whatever valid card GUID (af14ca09-a83d-4185-afa0-bc38a31dbf82
is your custom zone border card) in the game you want I will just use this
card to visualize the borders and make the necessary adjustments. I would
want the center of the card to be placed on the zone border. In reality two
or three would do but in reality I just need to figure out where your wall
card centering code is.

**Comment by sIKE23** [comment]
Played with the quite a bit this evening. Focused in the 6x4 Double Westlock board. The logic for the start corner is off for some reason. Start OCTGN in dev mode and then in the dev cosole paste in and run this function:

```python
def boardtester():
	boardSet = "GameBoard2.jpg"
	defineRectangularMap(6,4,170)
	for p in players:
		remoteCall(p, "setGameBoard", [boardSet])
	topleftcard = table.create("af14ca09-a83d-4185-afa0-bc38a31dbf82",-530,-415)
	bottomleftcard = table.create("af14ca09-a83d-4185-afa0-bc38a31dbf82",-530,-245)
	toprightcard = table.create("af14ca09-a83d-4185-afa0-bc38a31dbf82",-360,-415)
	bottomrightcard = table.create("af14ca09-a83d-4185-afa0-bc38a31dbf82",-360,-245)
	mapDict = eval(getGlobalVariable('Map'))
	scipttopleftcard = table.create("af14ca09-a83d-4185-afa0-bc38a31dbf82",mapDict.get('x'),mapDict.get('y'))
```
I am not sure of what is going on but do have my suspicions.

**Comment by sIKE23** [comment]
ok I played with this quite a bit tonight and using my cards I was able to
determine that the zone size for the 6x4 Westlock board is 170. However
using that number the formula that you are using does not properly
calculate the top right corner of the board. on the X axis is about 10
pixels over the right edge of the board, and on the Y axis low from the
corner by about  20ish pixels. This means that the zones are not placed
properly on the board. To take a look at what I mean there is a very simple
function in the dev folder that lays out cards for the start zone. You can
fire up dev mode and see how things are supposed to layout and where the
game thinks the top left corner is.

**Comment by ACG8** [comment]
I think I see what the problem is. As you noted, the game expects an aspect ratio of 4:3. The two problem boards have ratios of 6:4 and 5:3, which are not compatible. Unless there is a way to change the width and height of the board as defined in the XML file for the two maps, I don't think we can get the zone logic working properly without writing in some very specific exceptions for those boards (to offset the zones). That is doable, but it means that we have to add exceptions every time we add a new map. I would rather have a general zone-assigning function that works with all maps. I'll see if I can figure out a way around this.

**Comment by sIKE23** [comment]
I don't know if you are on the octgn thread about turning off hoover on
certain cards but there was movement on it and we can always go back that
direction using cards for tiles..

**Comment by ACG8** [comment]
Did some more experimentation. It looks like OCTGN assigns a fixed region to be occupied by the board, and resizes the board image to fit that region (without distorting it). I tried shrinking the board down to 5x5 pixels, and it expanded to fit in the region. I don't know if the size of the board region is defined somewhere (it isn't in the definition file; the width and length there appear to only affect the starting view of the board), or if it is a feature of OCTGN. 

Using cards for tiles for the 6x4 board is very ugly (since we have to chop it up into lots of tiles and assign each a separate GUID). Not sure how to proceed. I'll post an issue on OCTGN.

**Comment by sIKE23** [comment]
Brine talked about adding a board size to the api, I think that would
resolve that problem, no?

**Comment by ACG8** [comment]
Yes, I think that would probably do it.

**Comment by sIKE23** [comment]
This is done for the 4:3 perspective, waiting on an OCTGN fix for the others, #270 will carry the torch forward.
